rewriter/Flatten.cc

```diff
@@ -221,57 +212,65 @@ class FlattenWalk {
-            return rhs;
-        }
-
-        auto exprs = popCurMethodDefs();
+        auto currentMethodDefs = popCurMethodDefs();
         // this adds all the methods at the appropriate 'staticness' level
 
         int highestLevel = 0;
-        for (int i = 0; i < exprs.size(); i++) {
-            auto &expr = exprs[i];
+        for (int i = 0; i < currentMethodDefs.size(); i++) {
+            auto &expr = currentMethodDefs[i];
             if (highestLevel < expr.staticLevel) {
                 highestLevel = expr.staticLevel;
             }
             // we need to make sure that we keep sends with their attached methods, so fix that up here
             if (i > 0) {
-                auto send = ast::cast_tree<ast::Send>(exprs[i - 1].expr.get());
+                auto send = ast::cast_tree<ast::Send>(currentMethodDefs[i - 1].expr.get());
                 if (send != nullptr && send->fun == core::Names::sig()) {
-                    exprs[i - 1].staticLevel = expr.staticLevel;
+                    currentMethodDefs[i - 1].staticLevel = expr.staticLevel;
                 }
             }
         }
 
         // these will store the bodies of the `class << self` blocks we create at the end
-        vector<ast::ClassDef::RHS_store> nestedBlocks;
+        vector<ast::ClassDef::RHS_store> nestedClassBodies;
         for (int level = 2; level <= highestLevel; level++) {
-            nestedBlocks.emplace_back();
+            nestedClassBodies.emplace_back();
         }
 
         // this vector contains all the possible RHS target locations that we might move to
         vector<ast::ClassDef::RHS_store *> targets;
-        // 0 and 1 both go into the class itself
-        targets.emplace_back(&rhs);
+        vector<vector<unique_ptr<ast::Expression>>> expressionsToBePutInTargets;
+        // 0 and 1 both go into the class itself, but we want to make sure we do it while preserving
+        // order
         targets.emplace_back(&rhs);
+        targets.emplace_back(nullptr);
         // 2 and up go into the to-be-created `class << self` blocks
-        for (auto &tgt : nestedBlocks) {
+        for (auto &tgt : nestedClassBodies) {
             targets.emplace_back(&tgt);
         }
+        expressionsToBePutInTargets.resize(targets.size());
 
         // move everything to its appropriate target
-        for (auto &expr : exprs) {
+        for (auto &expr : currentMethodDefs) {
             if (auto methodDef = ast::cast_tree<ast::MethodDef>(expr.expr.get())) {
                 methodDef->setIsSelf(expr.staticLevel > 0);
             }
-            targets[expr.staticLevel]->emplace_back(std::move(expr.expr));
+            auto lvl = expr.staticLevel == 1 ? 0 : expr.staticLevel;
+            expressionsToBePutInTargets[lvl].emplace_back(std::move(expr.expr));
+        }
+
+        {
+            int idx = -1;
+            for (auto &target : targets) {
+                idx += 1;
+                if (idx != 1) {
```

Comment:
What's the significance of the entry at `1` here? Could you add a comment to clarify this?

Reply:
For one-level of `class << self` nesting, we flatten those methods into `self.` one static level up.

I added some comments and an ENFORCE, but TL;DR the line you've commented on can only really be understood in the context of this whole function.